src/Dialect/ONNX/DialectBuilder.cpp

```diff
@@ -376,6 +376,34 @@ Value OnnxBuilder::shape(
       toTensor(outputType), toTensor(input), endAttr, startAttr);
 }
 
+// Get the shape of an input and perform a permutation on it. Perm values are
+// in the range [0, rank(input)). Type is inferred. Operation get the dimensions
+// using onnx.dim and use onnx.concat to place the right value at the right
+// position.
+Value OnnxBuilder::shape(Value input, mlir::ArrayRef<int64_t> perm) const {
+  ShapedType inputType = mlir::cast<ShapedType>(input.getType());
+  int64_t inputRank = inputType.getRank();
+  auto inputShape = inputType.getShape();
+  int64_t permRank = perm.size();
+  bool allStatic = true;
+  llvm::SmallVector<Value, 4> permutedDims;
+  llvm::SmallVector<int64_t, 4> permutedShapes;
+  for (int64_t p = 0; p < permRank; ++p) {
+    int64_t d = perm[p];
+    assert(
+        d >= 0 && d < inputRank && "perm values expected in [0..rank(input))");
+    if (ShapedType::isDynamic(inputShape[d]))
+      allStatic = false;
+    permutedDims.emplace_back(dim(input, d));
```

Comment:
When the shape is static, these DimOp are still created. I know that they will be removed later by canonicalization, but they are quite redundant. We can have a light check like:
```c
bool isDynamic = llvm::any_of(inputShape, [](int64_t d) {return ShapedType::isDynamic(d);});
```

Then in the loop:
```
for (int64_t p = 0; p < permRank; ++p) {
   ...
   if (isDynamic)
      permutedDims.emplace_back(dim(input, d));
   ...
}

```